Add a representer for AnsibleUnsafeBytes

s-hertel opened the pull request:
##### SUMMARY
Fixes #62562

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/parsing/yaml/dumper.py

s-hertel committed:
Add a representer for AnsibleUnsafeBytes

willthames commented:
Thanks @s-hertel, much appreciated

s-hertel committed:
changelog

samdoran committed:
Add unit tests

samdoran committed:
Change test text slightly

abadger commented on test/units/parsing/yaml/test_dumper.py:
```diff
@@ -59,3 +60,40 @@ def test(self):
+
+        data_from_yaml = loader.get_single_data()
+
+        self.assertEqual(n_text, data_from_yaml)
```
I'm not sure this should work (that may be a "this shouldn't work in the future even if it works now")  I think AnsibleUnsafeText should convert any string it's given into a text string, ie: unicode type on python2 and str type on python3.

This may be the case now but the self.assertEqual is just not picking it up.  For each of your three test cases, try putting non-ascii characters into them.

(Note that the bytes one will needs a special incantation so it works on python3:
``` python
b_text = u'some byte string café'.encode('utf-8')
```

samdoran replied:
Non-ASCII characters in Python 3 pass tests:
```
 0.1s test_bytes: data_from_yaml=b'some byt\xc3\xa9s',
      type(data_from_yaml)=<class 'bytes'>

 0.1s test_native_text: data_from_yaml='some nativé text',
      type(data_from_yaml)=<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>

 0.1s test_unicode: data_from_yaml='some unicodé',
      type(data_from_yaml)=<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>
```

Python 2 fails with non-ASCII characters:
```
# pytest output
self = <units.parsing.yaml.test_dumper.TestAnsibleDumper testMethod=test_bytes>
>       self.assertEqual(b_text, data_from_yaml)
E       AssertionError: 'some byt\xc3\xa9s' != u'some byt\xe9s'

self = <units.parsing.yaml.test_dumper.TestAnsibleDumper testMethod=test_native_text>
>       f = AnsibleUnsafeText(n_text)
E       UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 10: ordinal not in range(128)

# some q output
 0.1s test_bytes: data_from_yaml=u'some byt\xe9s',
      type(data_from_yaml)=<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>

 0.1s test_unicode: data_from_yaml=u'some unicod\xe9',
      type(data_from_yaml)=<class 'ansible.parsing.yaml.objects.AnsibleUnicode'>
```

abadger replied:
Yep... Reading the Unsafe code, AnsibleUnsafeText and AnsibleUnsafeBytes are just light wrappers around the equivalent python types.  Since the equivalent python types are fragile when mixing types, this is fragile as well.  I think we need to consider it invalid to pass Text type to AnsibleUnsafeBytes and invalid to pass Bytes type to AnsibleUnsafeText.  By extension, it's also invalid to pass Native strings to either AnsibleUnsafeText or AnsibleUnsafeBytes because that will be wrong on one of python2 or python3.

samdoran committed:
Update tests

Remove native string test until we have time to evaluate how this the function should work
Add non-ASCII characters to test cases

samdoran committed:
Better French

samdoran committed:
Compare to the string on Python 2

Add a note about this behavior

abadger commented on test/units/parsing/yaml/test_dumper.py:
```diff
@@ -59,3 +61,36 @@ def test(self):
+            # pyyaml on Python 2 returns unicode and not bytes when given bytes with non-ASCII characters.
+            # Changes need to be made to the methods in Ansible to coerce the strings to behave consistently
+            # on Python 2 and Python 3.
+            # This is a workaround for the current behavior.
```
Little more accurate version of this comment
```
            # https://pyyaml.org/wiki/PyYAMLDocumentation#string-conversion-python-2-only
            # pyyaml on Python 2 can return either unicode or bytes when given byte strings.
            # We normalize that to always return unicode on Python2 as that's right most of the
            # time.  However, this means byte strings can round trip through yaml on Python3 but
            # not on Python2.  To make this code work the same on Python2 and Python3 (we want
            # the Python3 behaviour) we need to change the methods in Ansible to:
            # (1) Let byte strings pass through yaml without being converted on Python2
            # (2) Convert byte strings to text strings before being given to pyyaml  (Without this,
            #       strings would end up as byte strings most of the time which would mostly be wrong)
            # In practice, we mostly read bytes in from files and then pass that to pyyaml, for which
            # the present behavior is correct.
            # This is a workaround for the current behavior.
```

samdoran replied:
Thanks!

samdoran committed:
Update comment in test

s-hertel commented:
Thanks for the reviews and @sdoran for writing tests!
